File: codenvy-hosted-subscription-saas/src/main/java/com/codenvy/api/subscription/saas/server/billing/BillingService.java
/*
 * CODENVY CONFIDENTIAL
 * __________________
 * 
 *  [2012] - [2015] Codenvy, S.A. 
 *  All Rights Reserved.
 * 
 * NOTICE:  All information contained herein is, and remains
 * the property of Codenvy S.A. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Codenvy S.A.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Codenvy S.A..
 */
package com.codenvy.api.subscription.saas.server.billing;


import com.codenvy.api.subscription.saas.shared.dto.AccountResources;
import com.codenvy.api.subscription.saas.shared.dto.Invoice;
import com.codenvy.api.subscription.saas.shared.dto.Resources;

import org.eclipse.che.api.core.NotFoundException;
import org.eclipse.che.api.core.ServerException;

import java.util.List;

/**
 * Provide access to existed invoices and generate new invoices for the given period.
 *
 * @author Sergii Kabashniuk
 */
public interface BillingService {
    /**
     * Generate invoices for the given period of time.
     *
     * @param from
     *         beginning of period.
     * @param till
     *         end of period.
     * @return number of generated invoices
     * @throws ServerException
     */
    int generateInvoices(long from, long till) throws ServerException;

    /**
     * Get list of invoices with given Payment state.
     *
     * @param filter
     *         filter condition
     * @return - list of invoices with given Payment state.
     * @throws ServerException
     */
    List<Invoice> getInvoices(InvoiceFilter filter) throws ServerException;

    /**
     * Get invoice by id.
     *
     * @param id
     *         id of invoice
     * @return invoice
     * @throws ServerException
     */
    Invoice getInvoice(long id) throws ServerException, NotFoundException;

    /**
     * Change payment state of invoice.
     *
     * @param invoiceId
     *         id of invoice.
     * @param state
     *         next state of invoice.
     * @param creditCard
     *         credit card id. Null if no credit card was involved during operation with invoice.
     * @throws ServerException
     */
    void setPaymentState(long invoiceId, PaymentState state, String creditCard) throws ServerException;


    /**
     * Mark invoice as sent to user.
     *
     * @param invoiceId
     *         Id of invoice.
     * @throws ServerException
     */
    void markInvoiceAsSent(long invoiceId) throws ServerException;

    /**
     * Add prepaid GB*h
     *
     * @param accountId
     *         id of account for whom will be added prepaid GB*h
     * @param amount
     *         prepaid GB*h
     * @param from
     *         period when prepaid GB*h is active
     * @param till
     *         period when prepaid GB*h is active
     */
    void addSubscription(String accountId, double amount, long from, long till) throws ServerException;

    /**
     * Deactivate prepaid GB*h for account
     *
     * @param accountId
     *         id of account for whom will be deactivated prepaid GB*h
     * @param till
     *         period when prepaid GB*h is inactive
     */
    void removeSubscription(String accountId, long till) throws ServerException;

    /**
     * Get total used resources by given period.
     *
     * @return resources related to given account by given period
     */
    Resources getEstimatedUsage(long from, long till) throws ServerException;

    /**
     * Get resources related to accounts by given period.
     *
     * @return resources related to accounts by given period
     */
    List<AccountResources> getEstimatedUsageByAccount(ResourcesFilter resourcesFilter) throws ServerException;

    /**
     * Checks availability of resources for account by given period.
     *
     * @param accountId
     *         id of account
     * @param from
     *         begin of period
     * @param till
     *         end of period
     * @return if account has available resources {@code true} else {@code false}
     * @throws ServerException
     */
    boolean hasAvailableResources(String accountId, Long from, Long till) throws ServerException;

    /**
     * Get provided free resources for account by given period.
     *
     * @param accountId
     *         id of account
     * @param from
     *         begin of period
     * @param till
     *         end of period
     * @return number of provided free resources
     * @throws ServerException
     */
    double getProvidedFreeResources(String accountId, Long from, Long till) throws ServerException;

}

File: dashboard/src/app/billing/billing.service.ts
/*
 *  [2015] - [2017] Codenvy, S.A.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Codenvy S.A. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to Codenvy S.A.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Codenvy S.A..
 */
'use strict';

import {CodenvyPayment, ICreditCard} from '../../components/api/codenvy-payment.factory';

export class BillingService {
  $log: ng.ILogService;
  $q: ng.IQService;
  cheAPI: any;
  codenvyPayment: CodenvyPayment;

  /**
   * @ngInject for Dependency injection
   */
  constructor ($q: ng.IQService, $log: ng.ILogService,  cheAPI: any,  codenvyPayment: CodenvyPayment) {
    this.$log = $log;
    this.$q = $q;
    this.cheAPI = cheAPI;
    this.codenvyPayment = codenvyPayment;
  }

  /**
   * Fetches credit card.
   *
   * @param {string} accountId
   *
   * @returns {IPromise<any>}
   */
  fetchCreditCard(accountId: string): ng.IPromise<any> {
    let defer = this.$q.defer();

    this.codenvyPayment.fetchAllCreditCards(accountId).then(() => {
      let creditCard = this.getCreditCard(accountId);
      defer.resolve(creditCard);
    }, (error: any) => {
      if (error.status === 304) {
        let creditCard = this.getCreditCard(accountId);
        defer.resolve(creditCard);
      } else {
        this.$log.error(error);
        defer.reject(error);
      }
    });

    return defer.promise;
  }

  /**
   * Returns a credit card
   *
   * @param {string} accountId
   * @return {ICreditCard}
   */
  getCreditCard(accountId: string): ICreditCard {
    let creditCards = this.codenvyPayment.getCreditCards(accountId) || [];
    return creditCards[0] || null;
  }

  /**
   * Deletes existing credit card.
   *
   * @param {string} accountId
   * @param {string} token
   *
   * @return {ng.IPromise<any>}
   */
  removeCreditCard(accountId: string, token: string): ng.IPromise<any> {
    return this.codenvyPayment.removeCreditCard(accountId, token).then(() => {
      angular.noop();
    }, (error: any) => {
      this.$log.error(error);
    });
  }

  /**
   * Adds new credit card.
   *
   * @param {string} accountId
   * @param {ICreditCard} creditCard
   *
   * @return {ng.IPromise<any>}
   */
  addCreditCard(accountId: string, creditCard: ICreditCard): ng.IPromise<any> {
    return this.codenvyPayment.addCreditCard(accountId, creditCard);
  }

  /**
   * Updates an existing credit card.
   *
   * @param {string} accountId
   * @param {ICreditCard} creditCard
   *
   * @return {ng.IPromise<any>}
   */
  updateCreditCard(accountId: string, creditCard: ICreditCard): ng.IPromise<any> {
    return this.codenvyPayment.updateCreditCard(accountId, creditCard);
  }

}
